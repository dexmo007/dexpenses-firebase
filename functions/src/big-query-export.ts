import * as functions from 'firebase-functions';
import { BigQuery, Geography } from '@google-cloud/bigquery';
import { Timestamp } from '@google-cloud/firestore';
import { Change } from 'firebase-functions';
import { DocumentSnapshot } from 'firebase-functions/lib/providers/firestore';

function extractLocationGeography(receipt: any): Geography | undefined {
  const place = receipt.place;
  if (!place || !place.geometry || !place.geometry.location) {
    return undefined;
  }
  const { lng, lat } = place.geometry.location;
  if (!lng || !lat) {
    return undefined;
  }
  return BigQuery.geography(`POINT(${lng} ${lat})`);
}

interface BigQueryRow {
  bq_inserted_at: Date;
  bq_op: 'D' | 'I';
}

function fromSnapshotChange(snap: Change<DocumentSnapshot>): BigQueryRow {
  return {
    bq_inserted_at: (
      snap.after.updateTime ||
      snap.after.createTime ||
      Timestamp.now()
    ).toDate(),
    bq_op: !snap.after.exists ? 'D' : 'I',
  };
}

interface BigQueryReceipt extends BigQueryRow {
  user_id: string;
  receipt_id: string;
  tags?: string[];
  amount?: number;
  currency?: string;
  payment_method?: string;
  location?: Geography;
  timestamp?: Date;
  header?: string[];
}

export const receiptsToBigQuery = functions.firestore
  .document('receiptsByUser/{userId}/receipts/{receiptId}')
  .onWrite(async (snap, context) => {
    const bigQuery = new BigQuery();
    const userId = context.params.userId;
    const receiptId = context.params.receiptId;
    let row: BigQueryReceipt = {
      ...fromSnapshotChange(snap),
      user_id: userId,
      receipt_id: receiptId,
    };
    if (snap.after.exists) {
      const data = snap.after.data() as any;
      const receipt = data.result.data;
      if (!receipt) {
        console.log(`No receipt data for ${userId}/${receiptId}`);
        return;
      }
      row = {
        ...row,
        tags: data.tags || [],
        amount: (receipt.amount || {}).value,
        currency: (receipt.amount || {}).currency,
        payment_method: receipt.paymentMethod,
        location: extractLocationGeography(receipt),
        timestamp: receipt.timestamp ? receipt.timestamp.toDate() : null,
        header: receipt.header || [],
      };
    }

    try {
      await bigQuery
        .dataset('dexpenses_bi')
        .table('receipts')
        .insert(row);
    } catch (e) {
      console.error(JSON.stringify(e));
      throw e;
    }
  });

interface BigQueryRecurringPayment extends BigQueryRow {
  payment_id: string;
  user_id: string;
  name?: string;
  description?: string;
  period?: 'Monthly' | 'Quarterly' | 'Yearly';
  c_period?: number;
  amount?: number;
  currency?: string;
  payment_method?: string;
  due?: string;
}

function resolvePaymentPeriod(period?: string): number | undefined {
  if (!period) {
    return undefined;
  }
  switch (period) {
    case 'Monthly':
      return 1;
    case 'Quarterly':
      return 4;
    case 'Yearly':
      return 12;
    default:
      return undefined;
  }
}

export const recurringPaymentsToBigQuery = functions.firestore
  .document('recurringPaymentsByUser/{userId}/recurringPayments/{paymentId}')
  .onWrite(async (snap, context) => {
    const bigQuery = new BigQuery();
    const userId = context.params.userId;
    const paymentId = context.params.paymentId;
    let row: BigQueryRecurringPayment = {
      ...fromSnapshotChange(snap),
      user_id: userId,
      payment_id: paymentId,
    };
    if (snap.after.exists) {
      const data = snap.after.data() as any;
      row = {
        ...row,
        name: data.name,
        description: data.description,
        amount: (data.amount || {}).value,
        currency: (data.amount || {}).currency,
        payment_method: data.paymentMethod,
        period: data.period,
        c_period: resolvePaymentPeriod(data.period),
      };
    }

    try {
      await bigQuery
        .dataset('dexpenses_bi')
        .table('recurring_payments')
        .insert(row);
    } catch (e) {
      console.error(JSON.stringify(e));
      throw e;
    }
  });
