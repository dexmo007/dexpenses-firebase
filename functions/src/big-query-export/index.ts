export { receiptsToBigQuery } from './receipts';
export { recurringPaymentsToBigQuery } from './recurring-payments';
