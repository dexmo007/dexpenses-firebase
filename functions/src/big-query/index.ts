import * as functions from 'firebase-functions';
import { BigQuery } from '@google-cloud/bigquery';

export const aggregateTotal = functions.https.onCall(async (data, context) => {
  const query = `SELECT round(sum(amount), 2) as total
  FROM \`dexpenses-207219.dexpenses_bi.actual_receipts\`
  WHERE user_id = @user_id
  AND timestamp >= @start
  AND timestamp <= @end`;
  const [[result]] = await new BigQuery().query({
    query,
    params: {
      user_id: 'test',
      start: data.start,
      end: data.end || new Date(),
    },
  });
  return { value: result.total };
});

const PERIODS_TO_FIELD = {
  hourly: 'hour',
  daily: 'day',
  monthly: 'month',
  yearly: 'year',
};
const VALID_PERIODS = Object.keys(PERIODS_TO_FIELD);

export const aggregateTotalOverTimePeriod = functions.https.onCall(
  async (data, context) => {
    const { period, start, end } = data;
    if (!period || !VALID_PERIODS.includes(period)) {
      throw new functions.https.HttpsError(
        'failed-precondition',
        `invalid period: must be one of ${VALID_PERIODS}`
      );
    }
    const params = {
      user_id: 'test',
      start,
      end,
    };
    const aggregations = VALID_PERIODS.slice(VALID_PERIODS.indexOf(period))
      .reverse()
      .map((a) => PERIODS_TO_FIELD[a]);
    const extractorAs = (a: string) => `extract(${a} from timestamp) as ${a}`;
    const asc = (a: string) => `${a} asc`;
    const query = `SELECT ${aggregations
      .map(extractorAs)
      .join(', ')}, round(sum(amount), 2) as total
    FROM \`dexpenses-207219.dexpenses_bi.actual_receipts\`
    where user_id = @user_id
    and timestamp >= @start
    and timestamp <= @end
    group by ${aggregations.join(', ')}
    order by ${aggregations.map(asc).join(', ')}`;
    const [result] = await new BigQuery().query({
      query,
      params,
    });
    return result;
  }
);
