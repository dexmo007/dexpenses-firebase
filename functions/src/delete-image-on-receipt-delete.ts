import * as admin from 'firebase-admin';
import * as functions from 'firebase-functions';

export const deleteImageOnReceiptDelete = functions.firestore
  .document('receiptsByUser/{userId}/receipts/{receiptId}')
  .onDelete((snap, context) => {
    return admin
      .storage()
      .bucket('dexpenses-207219.appspot.com')
      .file(`images/${context.params.userId}/${context.params.receiptId}`)
      .delete();
  });
