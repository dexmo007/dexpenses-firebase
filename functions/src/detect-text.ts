import { ImageAnnotatorClient } from '@google-cloud/vision';
import * as admin from 'firebase-admin';
import * as functions from 'firebase-functions';
import { basename, dirname } from 'path';
import { PubSub } from '@google-cloud/pubsub';

interface TextDetectionResult {
  state: 'error' | 'no-text' | 'success';
  error?: Status;
  text?: string;
}

export interface TextDetectionResultMessage {
  text: string;
  userId: string;
  fileName: string;
}

interface AnnotateImageResponse {
  error?: Status;
  fullTextAnnotation?: TextAnnotation;
}
interface Status {
  code?: number;
  details?: Array<Record<string, any>>;
  message?: string;
}
interface TextAnnotation {
  text?: string;
}

function processVisionResult(
  result: AnnotateImageResponse
): TextDetectionResult {
  if (result.error) {
    return {
      state: 'error',
      error: result.error,
    };
  }
  if (!result.fullTextAnnotation || !result.fullTextAnnotation.text) {
    return {
      state: 'no-text',
    };
  }
  return {
    state: 'success',
    text: result.fullTextAnnotation.text,
  };
}

export async function runTextDetection(gsUrl: string) {
  const client = new ImageAnnotatorClient();
  const [annotateImageResponse] = await client.textDetection(gsUrl);
  return processVisionResult(annotateImageResponse);
}

export const detectText = functions.storage
  .object()
  .onFinalize(async (object) => {
    const userId = basename(dirname(object.name!));
    const fileName = basename(object.name!);
    const result = await runTextDetection(
      `gs://${object.bucket}/${object.name}`
    );
    if (result.state !== 'success') {
      return admin
        .firestore()
        .collection('detectionErrorsByUser')
        .doc(userId)
        .collection('detectionErrors')
        .doc(fileName)
        .set(result);
    }
    const message: TextDetectionResultMessage = {
      text: result.text!,
      userId,
      fileName,
    };
    return new PubSub()
      .topic('extraction')
      .publish(Buffer.from(JSON.stringify(message)));
  });
