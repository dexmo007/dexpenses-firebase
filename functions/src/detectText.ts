import * as vision from '@google-cloud/vision';
import * as admin from 'firebase-admin';
import * as functions from 'firebase-functions';
import * as path from 'path';
import { ReceiptResult } from './extract/receipt';

const BAD_IMAGE = 3;

export const detectText = functions.storage
  .object()
  .onFinalize(async (object) => {
    if (!object.name) {
      return;
    }
    const userId = path.basename(path.dirname(object.name));
    const fileName = path.basename(object.name);
    console.log('OCR', userId, fileName);
    const client = new vision.ImageAnnotatorClient();
    const [result] = await client.documentTextDetection(
      `gs://${object.bucket}/${object.name}`
    );
    if (result.error) {
      await admin
        .storage()
        .bucket()
        .file(object.name)
        .delete();
      let receiptResult: ReceiptResult;
      if (result.error.code === BAD_IMAGE) {
        receiptResult = {
          state: 'bad-image',
        };
      } else {
        receiptResult = {
          state: 'error',
          error: result.error,
        };
      }
      return admin
        .firestore()
        .collection('receiptsByUser')
        .doc(userId)
        .collection('receipts')
        .doc(fileName)
        .set(receiptResult, {
          merge: true,
        });
    }
    const fullTextAnnotation = result.fullTextAnnotation;
    if (!fullTextAnnotation) {
      return admin
        .firestore()
        .collection('receiptsByUser')
        .doc(userId)
        .collection('receipts')
        .doc(fileName)
        .set(
          {
            state: 'no-text',
          } as ReceiptResult,
          {
            merge: true,
          }
        );
    }
    return admin
      .firestore()
      .collection('receiptTextsByUser')
      .doc(userId)
      .collection('receiptTexts')
      .doc(fileName)
      .set({
        text: fullTextAnnotation.text,
      });
  });
