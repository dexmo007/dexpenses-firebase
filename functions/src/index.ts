import * as vision from "@google-cloud/vision";
import * as admin from "firebase-admin";
import * as functions from "firebase-functions";
import * as path from "path";
import extractorPipeline from "./extract/pipeline";

admin.initializeApp();

export const detectText = functions.storage.object().onFinalize(async (object) => {
  if (!object.name) {
    return;
  }
  const userId = path.basename(path.dirname(object.name));
  const fileName = path.basename(object.name);
  console.log("OCR", userId, fileName);
  const client = new vision.ImageAnnotatorClient();
  const [result] = await client.documentTextDetection(`gs://${object.bucket}/${object.name}`);
  const fullTextAnnotation = result.fullTextAnnotation;
  return admin.firestore()
    .collection("receiptTextsByUser").doc(userId).collection("receiptTexts")
    .doc(fileName)
    .set(
      {
        text: fullTextAnnotation.text
      }
    );
});

export const analyseReceiptText = functions.firestore.document("receiptTextsByUser/{userId}/receiptTexts/{receiptId}")
  .onCreate((snap, context) => {
    const data = snap.data();
    const result = extractorPipeline((data || {}).text);
    return admin.firestore()
      .collection("receiptsByUser").doc(context.params.userId)
      .collection("receipts").doc(context.params.receiptId)
      .set({
        result
      }, {
          merge: true,
        });
  });
