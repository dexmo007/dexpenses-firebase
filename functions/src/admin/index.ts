import { extname } from 'path';
import * as admin from 'firebase-admin';
import * as functions from 'firebase-functions';
import * as Octokit from '@octokit/rest';
import { onAuthorizedCall, anyOf } from '../https';
import { validateNotBlank, validateRequired } from '../validation';
import { runTextDetection } from '../detect-text';
import { PaymentMethod, paymentMethods } from '@dexpenses/core';
import { parseGSUrl } from '../util/gsutil';

export const testImageBucket = 'dexpenses-207219-test-images';
export const testImageUploadBucket = 'dexpenses-207219-test-image-upload';

export interface TestDataInfo {
  category: string;
  cityCode: string;
  name: string;
  classifier?: string;
  paymentMethod?: PaymentMethod;
}

function validateTestDataInfo(
  info: Partial<TestDataInfo> & { paymentMethod?: string }
): TestDataInfo {
  validateRequired(info);
  validateNotBlank(info.category, 'category');
  validateNotBlank(info.cityCode, 'cityCode');
  validateNotBlank(info.name, 'name');
  if (
    info.paymentMethod &&
    !paymentMethods.includes(
      info.paymentMethod.trim().toUpperCase() as PaymentMethod
    )
  ) {
    throw new functions.https.HttpsError(
      'failed-precondition',
      'invalid payment method'
    );
  }
  return info as TestDataInfo;
}

export function buildIdentifier(info: TestDataInfo): string {
  return (
    info.category +
    '/' +
    [info.cityCode, info.name, info.classifier, info.paymentMethod]
      .map((s) => s && s.trim())
      .filter((p) => !!p)
      .map((s) => s!.toLowerCase().replace(/\s+/g, '-'))
      .join('-')
  );
}

export const addTestDataFile = onAuthorizedCall(anyOf('contributor'))(
  async (data, context) => {
    validateNotBlank(data.content, 'content');
    validateNotBlank(data.path, 'path');
    const info = validateTestDataInfo(data);

    const auth = functions.config().github.bot.key;
    const octokit = new Octokit({
      auth,
    });
    const owner = 'dexpenses';
    const repo = 'dexpenses-extract';
    const identifier = buildIdentifier(info);
    await octokit.repos.createOrUpdateFile({
      owner,
      repo,
      content: Buffer.from(data.content).toString('base64'),
      path: `test/data/${identifier}.txt.inactive`,
      message: `[skip ci] ðŸ¤– Added ${identifier} test file`,
    });
    const imageUrl = `https://firebasestorage.googleapis.com/v0/b/${testImageBucket}/o/${encodeURIComponent(
      identifier
    )}${extname(data.path)}?alt=media`;
    await octokit.issues.create({
      owner,
      repo,
      title: `Implement test receipt ${identifier}`,
      labels: ['enhancement', 'test-data'],
      body:
        'Receipt to implement:\n' +
        `![${identifier}](${imageUrl} "${identifier}")`,
    });
    console.log(`Added ${identifier} test file`);
    return { success: true };
  }
);

function validateGSUrl(uid: string, rawUrl: string) {
  const url = parseGSUrl(rawUrl);
  if (!url || !url.filename) {
    throw new functions.https.HttpsError(
      'failed-precondition',
      'invalid GS url'
    );
  }
  if (url.bucket !== testImageBucket) {
    throw new functions.https.HttpsError(
      'permission-denied',
      'permission-denied'
    );
  }
}

export const manualTextDetection = onAuthorizedCall(anyOf('contributor'))(
  async (data, context) => {
    validateNotBlank(data.url, 'url');
    validateGSUrl(context.auth.uid, data.url);
    return await runTextDetection(data.url);
  }
);

export const moveTestDataImage = onAuthorizedCall(anyOf('contributor'))(
  async (data, context) => {
    validateNotBlank(data.source, 'source');
    const info = validateTestDataInfo(data);
    const identifier = buildIdentifier(info);
    await admin
      .storage()
      .bucket(testImageUploadBucket)
      .file(data.source)
      .move(admin.storage().bucket(testImageBucket).file(`${identifier}${extname(data.source)}`));
    return { success: true };
  }
);
