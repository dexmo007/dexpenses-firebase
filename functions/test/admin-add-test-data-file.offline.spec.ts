import * as fbTest from 'firebase-functions-test';
import * as admin from 'firebase-admin';
import * as functions from 'firebase-functions';
import { firestore, firestoreDb } from './firebase-stubs';
import { TestDataInfo } from '../src/admin';

const test = fbTest();

jest.spyOn(admin, 'initializeApp');
jest.spyOn(functions, 'config').mockReturnValue({
  github: {
    bot: {
      key: 'key',
    },
  },
});

const mockCreateOrUpdateFile = jest.fn();
jest.mock('@octokit/rest', () =>
  jest.fn().mockImplementation(() => ({
    repos: {
      createOrUpdateFile: mockCreateOrUpdateFile,
    },
  }))
);

jest.spyOn(admin, 'firestore' as any, 'get').mockReturnValue(() => firestore);

let addTestDataFile;
beforeEach(() => {
  jest.clearAllMocks();
  firestoreDb['auth/test'] = { roles: 'contributor' };
  addTestDataFile = test.wrap(require('../src').addTestDataFile);
});

describe('admin functions/addTestDataFile', () => {
  it('should require "contributor" role', async () => {
    firestoreDb['auth/test'] = {};
    await expect(
      addTestDataFile(
        {},
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/permission-denied/);
  });
  const validData: { content: string } & TestDataInfo = {
    content: 'content',
    category: 'ec',
    cityCode: 'wob',
    name: 'name',
    paymentMethod: 'credit',
  };

  it('should throw error if content is missing or blank', async () => {
    await expect(
      addTestDataFile(
        { ...validData, content: undefined },
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/content/);
    await expect(
      addTestDataFile(
        { ...validData, content: '' },
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/content/);

    await expect(
      addTestDataFile(
        { ...validData, content: '  ' },
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/content/);
  });

  it('should throw error if category is missing', async () => {
    await expect(
      addTestDataFile(
        { ...validData, category: '  ' },
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/category/);
  });

  it('should throw error if cityCode is missing', async () => {
    await expect(
      addTestDataFile(
        { ...validData, cityCode: '  ' },
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/cityCode/);
  });

  it('should throw error if name is missing', async () => {
    await expect(
      addTestDataFile(
        { ...validData, name: '  ' },
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/name/);
  });

  it('should throw error if paymentMethod is missing', async () => {
    await expect(
      addTestDataFile(
        { ...validData, paymentMethod: '  ' },
        {
          auth: {
            uid: 'test',
          },
        }
      )
    ).rejects.toThrowError(/paymentMethod/);
  });

  it('should send correct request to GitHub', async () => {
    await expect(
      addTestDataFile(validData, { auth: { uid: 'test' } })
    ).resolves.toEqual({ success: true });

    expect(mockCreateOrUpdateFile.mock.calls[0]).toMatchSnapshot();
  });
});
