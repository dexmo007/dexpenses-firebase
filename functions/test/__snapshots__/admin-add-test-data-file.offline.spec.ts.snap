// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin functions/addTestDataFile should send correct request to GitHub 1`] = `
Array [
  Object {
    "content": "Y29udGVudA==",
    "message": "ðŸ¤– Added ec/wob-name-credit test file",
    "owner": "dexpenses",
    "path": "test/data/ec/wob-name-credit.txt.inactive",
    "repo": "dexpenses-extract",
  },
]
`;

exports[`admin functions/addTestDataFile should send correct request to GitHub 2`] = `
Array [
  Object {
    "body": "Receipt to implement:
![ec/wob-name-credit](https://firebasestorage.googleapis.com/v0/b/dexpenses-207219-test-images/o/ec%2Fwob-name-credit.jpg?alt=media \\"ec/wob-name-credit\\")",
    "labels": Array [
      "enhancement",
      "test-data",
    ],
    "owner": "dexpenses",
    "repo": "dexpenses-extract",
    "title": "Implement test receipt ec/wob-name-credit",
  },
]
`;
